docker: sign container images pushed to GHCR with GitHub OIDC tokens

konidev20 opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
* This allows users to verify the authenticity and integrity of Restic Docker images before using them. 
* This provides users with a verifiable record of how, when and where an artifact was built.

Refer:
1. https://github.blog/security/supply-chain-security/safeguard-container-signing-capability-actions/
2. ~~https://github.com/sigstore/cosign-installer~~
3. [Generation of SLSA3+ provenance for container images](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/container#generating-provenance)

[Edit]
Changes made to the GitHub Workflow:
* ~~installing cosign v2.4.0~~
* ~~add new step to sign all the images generated during the build and push step~~
* the id-token of the GitHub Actions workflow will be used for image signing
* replace branch-based tagging with SHA-based tagging since, branch names are mutable, SLSA provenance requires immutable tagging
* use official SLSA framework Github Reusable workflow
[\Edit]

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Closes #4983 

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

MichaelEischer commented on .github/workflows/docker.yml:
```diff
@@ -31,6 +36,16 @@ jobs:
 
+      # Install the cosign tool except on PR
+      # https://github.com/sigstore/cosign-installer
+      - name: Install cosign
```
Why exactly do we have to manually install cosign? https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/container#generating-provenance looks like we basically just have to add the `provenance` step and are done with it?

konidev20 replied:
Yeah, I had 2 changes as part of this MR actually.
1. The container signing and 
2. The other was the SLSA provenance. 

I will remove the container signing part. That seems like a separate issue.

MichaelEischer commented on .github/workflows/docker.yml:
```diff
@@ -64,3 +80,37 @@ jobs:
+      id-token: write # for creating OIDC tokens for signing.
+      packages: write # for uploading attestations.
+    if: github.repository == 'restic/restic'
+    uses: slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@v2.0.0
```
Is there a downside to using the generator framework? I'd expect it to be less work than using `cosign` directly

konidev20 replied:
Yes, it's lesser work than the container signing. I will be creating a separate Issue for that. 

In terms of security I reckon both are required, that's why I raised it in a single PR. 

Container Signing would be required for integrity and SLSA Provenance would be required for the origins of the build process.

konidev20 committed:
docker: add SLSA provenance to .github workflow

* the id-token of the GitHub Actions workflow will be used for image signing
* replace branch-based tagging with SHA-based tagging since, branch names are mutable, SLSA provenance requires immutable tagging
* use official SLSA framework Github Reusable workflow

docker: fix incorrect registry name in image output step

* use REGISTRY environment variable instead of IMAGE_REGISTRY

docker: revert change to remove branch tag

konidev20 commented:
@MichaelEischer need some more time to test. facing some issue during the slsa verification.

konidev20 committed:
docs: add instructions for verifying SLSA provenance of Docker images

docs: update the documentation

MichaelEischer commented on .github/workflows/docker.yml:
```diff
@@ -37,7 +42,8 @@ jobs:
           tags: |
-            type=ref,event=branch
+            type=sha
+            type=ref, event=branch
```
why is there an additional space between `type=ref` and `event=branch`?

konidev20 replied:
Fixed, this was introduced by mistake.

MichaelEischer commented on changelog/unreleased/issue-4983:
```diff
@@ -0,0 +1,8 @@
+Enhancement: add SLSA provenance to the Docker images
+
+Restic's Docker image build workflow now includes SLSA provenance generation. 
```
This only applies to the images on GHCR, not the images on Dockerhub.

konidev20 replied:
I am not sure how the build-and-push steps work to upload to Dockerhub. Let me see if I can make this a generic solution which would work for any repository, not limited to Dockerhub and GHCR.

MichaelEischer commented on .github/workflows/docker.yml:
```diff
@@ -19,6 +19,11 @@ jobs:
     permissions:
       contents: read
       packages: write
+      id-token: write # needed for signing the images with GitHub OIDC Token
```
The `id-token` is only necessary for the `provenance` step.

konidev20 replied:
Fixed, in my first approach, I had added the provenance step to the same job.

konidev20 committed:
docker: fix typos and permissions for jobs

konidev20 committed:
changelog: update to indicate change applies only for GHCR images.

konidev20 commented:
@MichaelEischer I guess you can do a final review and merge this. 

As for
> I am not sure how the build-and-push steps work to upload to Dockerhub. Let me see if I can make this a generic solution which would work for any repository, not limited to Dockerhub and GHCR.

I will pick it up in a separate issue.

MichaelEischer reviewed (approved):
LGTM. Thanks!

> I am not sure how the build-and-push steps work to upload to Dockerhub. Let me see if I can make this a generic solution which would work for any repository, not limited to Dockerhub and GHCR.

Currently the Dockerhub images are built locally, see https://github.com/restic/restic/blob/master/helpers/prepare-release/main.go for the corresponding script.

To get SLSA provenance for the DockerHub images, we have to allow the CI pipeline to push images to DockerHub. Then either the signature for the GHCR image can be reused, or the pipeline just duplicates the provenance generation step.
